// statable/dist/state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require("babel-runtime/helpers/createClass");

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var State = function () {
	function State(state, methods) {
		(0, _classCallCheck3.default)(this, State);

		this.state = state;
		if (methods) {
			for (var i in methods) {
				this[i] = methods[i].bind(this);
			}
		}
		this.subscriptions = [];
	}

	(0, _createClass3.default)(State, [{
		key: "setState",
		value: function setState(state) {
			var _this = this;

			Object.assign(this.state, state);
			this.subscriptions.forEach(function (subscription) {
				subscription(_this.state);
			});
		}
	}, {
		key: "subscribe",
		value: function subscribe(fn) {
			this.subscriptions.push(fn);
		}
	}, {
		key: "unsubscribe",
		value: function unsubscribe(fn) {
			var index = this.subscriptions.indexOf(fn);
			if (index > -1) {
				this.subscriptions.splice(index, 1);
			}
		}
	}]);
	return State;
}();

exports.default = State;
module.exports = exports["default"];
